### ISSUE TITLE
Angular: Build error "Cannot find name 'Buffer'" when using a bytes field.

### ISSUE THREAD

**Issue by timostamm** [stephenh/ts-proto#77]
First, thank you for this library, @stephenh. It has been a while since I looked around for a Typescript protobuf library and I was very happy to discover ts-proto last week. 

However, I did run into an issue with angular. The plugin generates the following code when a bytes field is present: 

```ts
interface WindowBase64 {
  atob(b64: string): string;
  btoa(bin: string): string;
}

const windowBase64 = (globalThis as unknown as WindowBase64);
const atob = windowBase64.atob || ((b64: string) => Buffer.from(b64, 'base64').toString('binary'));
const btoa = windowBase64.btoa || ((bin: string) => Buffer.from(bin, 'binary').toString('base64'));
```

The Angular compiler fails the build with the following error: 

```
error TS2591: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i @types/node` and then add `node` to the types field in your tsconfig.
const atob = windowBase64.atob || ((b64: string) => Buffer.from(b64, 'base64').toString('binary'));
                                                    ~~~~~~
```

I don't think I should follow the compiler`s  suggestion to install node's type definitions because that would pollute the application with types from node that it will never be able to use. 

The following replacement fixes the issue:

```ts
interface WindowBase64 {
  atob(b64: string): string;
  btoa(bin: string): string;
}
declare class Buffer {
  static from(str: string, encoding?: 'base64'|'binary'): Buffer;
  toString(encoding:'base64'|'binary'):string;
}
const windowBase64 = (typeof globalThis !== 'undefined' ? globalThis : this) as unknown as WindowBase64;
const atob = windowBase64.atob || ((b64: string) => Buffer.from(b64, 'base64').toString('binary'));
const btoa = windowBase64.btoa || ((bin: string) => Buffer.from(bin, 'binary').toString('base64'));
```

Jest tests pass and the angular compiler is happy. As an added bonus this adds support for Edge 18 and iOS 12 (they don't know `globalThis`). 

More compact version - maybe better suited for generated code:
```ts
declare const global: any;
const globalObj = typeof globalThis !== 'undefined' ? globalThis : (typeof global !== 'undefined' ? global : this) as any;
const atob = globalObj.atob || ((b64: string) => globalObj.Buffer.from(b64, 'base64').toString('binary'));
const btoa = globalObj.btoa || ((bin: string) => globalObj.Buffer.from(bin, 'binary').toString('base64'));
```

What do you think, are you interested in a PR?

**Comment by stephenh** [comment]
@timostamm hm, does `Buffer.from` actually _work_ in a browser? I get the `declare class Buffer` makes the TS compiler happy, but if that line is actually invoked, does it work? Do some of the browsers have `Buffer.from` implemented? 

I think I mentioned it else where, but I'd love to have an `integration/angular` directory with a super-simple version of your setup that shows an angular app using ts-proto output.

Could we start with that? I.e. a PR for the "toy angular app that uses ts-proto"?

**Comment by timostamm** [comment]
@stephenh, nodes' `Buffer` is built on top of the ECMA `Uint8Array` and not available in _any_ browser. 

A simple angular setup should not be a problem. A test suite with node and browser coverage on generated code on the other hand is not trivial at all. Let's start simple. I should be able to find some time in the coming week to setup a basic angular example.

**Comment by stephenh** [comment]
> not available in any browser

Okay, that makes sense. That's why I was thinking that the proposed fix of `declare class Buffer` is not really sufficient, because it will make the compile error go away (by making TS _think_ there is a `Buffer` class), but if the generated code ever calls that method, it won't actually work...

Like we probably need to change the generated code to realize "`Buffer` is not available" and use something else instead.

> in the coming week to setup a basic angular example

Sounds great.

**Comment by yiochen** [comment]
Can we include [buffer](https://www.npmjs.com/package/buffer) when generating for browser.

**Comment by stephenh** [comment]
I'm pretty sure this is fixed in v1.48.0. Basically prefixing `Buffer` with `globalThis.Buffer`. Let me know if not.
